plugins/Dispersion/Dispersion.cpp

```diff
@@ -0,0 +1,166 @@
+/*
+ * Dispersion.cpp
+ *
+ * Copyright (c) 2023 Lost Robot <r94231/at/gmail/dot/com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+#include "Dispersion.h"
+
+#include "embed.h"
+#include "plugin_export.h"
+
+namespace lmms
+{
+
+extern "C"
+{
+
+Plugin::Descriptor PLUGIN_EXPORT dispersion_plugin_descriptor =
+{
+	LMMS_STRINGIFY(PLUGIN_NAME),
+	"Dispersion",
+	QT_TRANSLATE_NOOP("PluginBrowser", "An allpass filter allowing for extremely high orders."),
+	"Lost Robot <r94231/at/gmail/dot/com>",
+	0x0100,
+	Plugin::Effect,
+	new PluginPixmapLoader("logo"),
+	nullptr,
+	nullptr,
+};
+
+}
+
+
+DispersionEffect::DispersionEffect(Model* parent, const Descriptor::SubPluginFeatures::Key* key) :
+	Effect(&dispersion_plugin_descriptor, parent, key),
+	m_dispersionControls(this),
+	m_sampleRate(Engine::audioEngine()->processingSampleRate()),
+	m_amountVal(0)
+{
+}
+
+
+bool DispersionEffect::processAudioBuffer(sampleFrame* buf, const fpp_t frames)
+{
+	if (!isEnabled() || !isRunning())
+	{
+		return false;
+	}
+
+	double outSum = 0.0;
+	const float d = dryLevel();
+	const float w = wetLevel();
+	
+	const int amount = m_dispersionControls.m_amountModel.value();
+	const float freq = m_dispersionControls.m_freqModel.value();
+	const float reso = m_dispersionControls.m_resoModel.value();
+	float feedback = m_dispersionControls.m_feedbackModel.value();
+	const bool dc = m_dispersionControls.m_dcModel.value();
+	
+	// Allpass coefficient calculation
+	const float w0 = (F_2PI / m_sampleRate) * freq;
+	const float a0 = 1 + (std::sin(w0) / (reso * 2.f));
+	float apCoeff1 = (1 - (a0 - 1)) / a0;
+	float apCoeff2 = (-2 * std::cos(w0)) / a0;
+	
+	float dcCoeff = 0.001 * (44100.f / m_sampleRate);
+	
+	if (amount != m_amountVal)
+	{
+		if (amount < m_amountVal)
+		{
+			// Flush filter buffers when they're no longer in use
+			for (int i = amount; i < m_amountVal; ++i)
+			{
+				m_apX0[i][0] = m_apX0[i][1] = m_apX1[i][0] = m_apX1[i][1] = 0;
+				m_apY0[i][0] = m_apY0[i][1] = m_apY1[i][0] = m_apY1[i][1] = 0;
+			}
+		}
+		m_amountVal = amount;
+	}
+	
+	if (amount == 0)
+	{
+		feedback = 0;
```

Comment:
I think that you need to reset the model as well, because if not you can get a click on the first sample if m_feedbackVal has some value